File: Part1_강의자료/Ch07_이분탐색/문제별코드/2343_기타레슨/2343.java
import java.util.Scanner;

class Main
{
    public static boolean isPossible(int[] lengths, int videoLength, int videoCount) {
        int currentLength = 0;
        int currentCount = 1;
        for (int len : lengths) {
            if (len > videoLength) return false;
            if (currentLength + len > videoLength) {
                if (++currentCount > videoCount) return false;
                currentLength = 0;
            }
            currentLength += len;
        }
        return true;
    }

    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int M = sc.nextInt();
        int[] lengths = new int[N];
        for (int i = 0; i < N; i++)
            lengths[i] = sc.nextInt();

        int l = 1, r = N * 10000, ans = -1;
        while (l <= r) {
            int m = (l + r) / 2;
            if (isPossible(lengths, m, M)) {
                ans = m;
                r = m - 1;
            }
            else l = m + 1;
        }
        System.out.println(ans);
    }
}

File: Part1_강의자료/Ch07_이분탐색/문제별코드/2343_기타레슨/2343.cpp
#include <stdio.h>
#include <vector>
using namespace std;

bool is_possible(vector<int> &lengths, int video_length, int video_count)
{
    int current_length = 0, current_count = 1;
    for (auto & len : lengths)
    {
        if (len > video_length)
            return false;
        if (current_length + len > video_length)
        {
            if (++current_count > video_count) 
                return false;
            current_length = 0;
        }
        current_length += len;
    }
    return true;
}

int main()
{
    int N, M;
    scanf("%d %d", &N, &M);
    
    vector<int> lengths(N);
    for (int i = 0; i < N; i++)
        scanf("%d", &lengths[i]);

    int l = 1, r = N * 10000, ans = -1;
    while (l <= r)
    {
        int mid = (l + r) / 2;
        if (is_possible(lengths, mid, M))
        {
            ans = mid;
            r = mid - 1;
        }
        else
            l = mid + 1;
    }
    printf("%d\n", ans);
    return 0;
}
